src/stub/core_impl/extended_mode.rs

```diff
@@ -26,6 +26,7 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
             }
             ExtendedMode::vAttach(cmd) => {
                 ops.attach(cmd.pid).handle_error()?;
+                res.write_str("S00")?;
```

Comment:
I'm not sure this is the approach I'd like to take here.

The [docs](https://sourceware.org/gdb/onlinedocs/gdb/Packets.html#index-vAttach-packet) for the `vAttach` packet clearly state that you should be able to return _any_ stop reason, so hard-coding `S00` doesn't seem like the best idea from a long-term correctness POV. Different implementations may very well have different ways of "attaching" to a running process (i.e: not just `S00`), and giving them the ability to report the right info off-the-bat (vs. forcing extra round-trips) seems better.

That's not to say I don't acknowledge and believe https://github.com/daniel5151/gdbstub/issues/126#issuecomment-1466282750, but I think it's reasonable to punt those sorts of considerations to the user.

* * *

For this initial implementation, lets have this arm return a `HandlerStatus::DeferredStopReason`, and update the docs of the `attach` handler with something along the lines of "once this method resolves, GDB will immediately request a stop reason. It is up to the implementation to ensure the stop reason is accurate".

Maybe in the future, we can think about ways to add some more "guard rails" here (e.g: limiting the scope of stop reasons, tweaking the attach handler to return a stop reason directly, etc...), but that'll require some intrusive changes and experimentations.

Reply:
How do I get `HandlerStatus::DeferredStopReason` to request a stop reason? If I return that from that function and don't output `S00` then GDB breaks due to not having a valid PID:

```
(gdb) att 5
Attaching to program: E:\Code\Xous\Core\target\riscv32imac-unknown-xous-elf\release\xous-ticktimer, process 5
[remote] Sending packet: $vAttach;5#3b
[remote] getpkt_or_notif_sane_1: Timed out.
[remote] getpkt_or_notif_sane_1: Timed out.
[remote] getpkt_or_notif_sane_1: Timed out.
Ignoring packet error, continuing...
[remote] packet_ok: Packet vAttach (attach) is supported
[remote] Sending packet: $qC#b4
[remote] Packet received: QC
/__w/riscv-none-elf-gcc-xpack/riscv-none-elf-gcc-xpack/build/win32-x64/sources/gdb-12.1/gdb/inferior.c:304: internal-error: find_inferior_pid: Assertion `pid != 0' failed.
A problem internal to GDB has been detected,
further debugging may prove unreliable.
Quit this debugging session? (y or n) Recursive internal problem.
```

Reply:
If I have it return `HandlerStatus::DeferredStopReason` then it doesn't work at all, even if I add back in `res.write_str("S00")?;`

Reply:
Ahh, indeed, I may have led you a little astray here. Rather than:

> once this method resolves, GDB will immediately request a stop reason. It is up to the implementation to ensure the stop reason is accurate

It should _really_ be:

> one this method resolves, GDB will expect to receive a stop reason corresponding to the attach event. The implementation is responsible for setting whatever internal flags it requires in order to immediately report an accurate stop reason after this method resolves.

That is to say: in much that same way that the `resume` method doesn't immediately return a stop reason - instead deferring the act of "waiting" for a stop reason to the event loop - so too should this method defer the act of reporting a stop reason to the event loop.

* * *

That said, this is a bit... weird.
I'm can't say i'm a _huge_ fan of this kind of unnecessary control flow cotorsion.

At the same time, `S00` also seems... off.

Here's what I think: can you lift the implementation of the `Base::QuestionMark` handler into a shared function (that can live in `base.rs`, as part of the `GdbStubImpl`) that also gets used in this handler? Oh, and make that function `#[inline(always)]` please.

After all - that is _effectively_ what we're doing here.

In the future, I'll think about how I want to properly support "custom" `?` and `vAttach` responses, but if we're hard-coding a "reasonable" response in `gdbstub` for the time being, I'd rather keep that response consistent between both packets.